### PR TITLE
Fix bug where L034 should ignore INSERT or "CREATE TABLE AS SELECT" with CTE

### LINKED ISSUE #4105: Rule 34 : SQlfluff doesn't ignore this rule for INSER INTO statements
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

Check this [SLACK conversation](https://sqlfluff.slack.com/archives/C01GPHR8J0G/p1669195341860829)
# Context : 
- Templater : Jinja
- Dialect : Snowflake

# What goes wrong ? : 
The rule 34 is meant to order the fields into a select statement : wildcards, simple targets then aggregations
It's a cool feature but this causes a problem when reordering fields during an INSERT INTO statement 

I took a look on your code and I saw that you already thought about something like this
[check code](https://github.com/sqlfluff/sqlfluff/blob/main/src/sqlfluff/rules/L034.py#L86-L88)

But Unfortunately this is not working as it should
Indeed whenever there is a `WITH` clause between the `INSERT INTO` and the `SELECT` 
SQLfluff struggles and doesn't always succeed in detecting it as an INSERT statement to ignore this rule

# How does SQLfluff performs ? : 
We have 2 cases : 
- INSERT INTO directly followed by a SELECT : 
```
INSERT INTO my_table
SELECT ( field1, field2 , ..) FROM ...
```
In this case SQLfluff can understand that it is an `INSERT` statement and it won't reorder the fields based on **L034** in the `SELECT`

- When there is a CTE in between (WITH clause) : 
```
INSERT INTO my_table
WITH my_cte AS (SELECT ...)
SELECT ( field1, field2 , ..) FROM ...
```
OR
```
INSERT INTO my_table
WITH my_cte AS (SELECT ...),
my_second_cte AS (SELECT ...) 
SELECT ( field1, field2 , ..) FROM ...
```

In this case **SQLFLUFF** struggles a lot and sometimes it detects it as an `INSERT` but many times it doesn't and it applied the rule **L034** which causes a lot of problems

### Expected Behaviour

Make sure SQLfluff can always detect an `INSERT` statement to not apply the rule L034

### Observed Behaviour

Whenever there is a `WITH` clause between the `INSERT INTO` and the `SELECT` 
SQLfluff struggles and doesn't always succeed in detecting it as an INSERT statement to ignore this rule

### How to reproduce

`sqlfluff fix` for :
```
INSERT INTO my_table
WITH my_cte AS (SELECT ...)
SELECT ( field1, field2 , ..) FROM ...
```
OR
```
INSERT INTO my_table
WITH my_cte AS (SELECT ...),
my_second_cte AS (SELECT ...) 
SELECT ( field1, field2 , ..) FROM ...
```

### Dialect

Snowflake

### Version

1.4.2

### Configuration

[sqlfluff]
dialect = snowflake
exclude_rules = L029, L031
fix_even_unparsable = False

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by barrywhart**
### Brief summary of the change made
Fixes #4105 

### Pull Request checklist
- [X] Please confirm you have completed any of the necessary steps below.

- Included test cases to demonstrate any code changes, which may be one or more of the following:
  - `.yml` rule test cases in `test/fixtures/rules/std_rule_cases`.
  - `.sql`/`.yml` parser test cases in `test/fixtures/dialects` (note YML files can be auto generated with `tox -e generate-fixture-yml`).
  - Full autofix test cases in `test/fixtures/linter/autofix`.
  - Other.
- Added appropriate documentation for the change.
- Created GitHub issues for any relevant followup/future enhancements if appropriate.

### CODE REVIEW

File: src/sqlfluff/core/templaters/slicers/tracer.py [hunk 1]
```diff
@@ -76,7 +76,7 @@ def trace(self, append_to_templated: str = "") -> JinjaTrace:
         trace_template = self.make_template(trace_template_str)
         trace_template_output = trace_template.render()
         # Split output by section. Each section has two possible formats.
-        trace_entries = list(regex.finditer(r"\0", trace_template_output))  # type: ignore[call-overload] # noqa: E501
+        trace_entries = list(regex.finditer(r"\0", trace_template_output))
```
Comment (barrywhart) [inline src/sqlfluff/core/templaters/slicers/tracer.py:79]: This recent mypy issue has now been fixed by the release of `types-regex` version 2022.10.31.1.

File: src/sqlfluff/rules/L034.py [hunk 2]
```diff
@@ -94,10 +94,24 @@ def _eval(self, context: RuleContext) -> EvalResultType:
             "insert_statement", "set_expression"
         ):
             return None
+        if (
+            len(context.parent_stack) >= 3
+            and context.parent_stack[-3].is_type("insert_statement", "set_expression")
```
Comment (tunetheweb) [inline src/sqlfluff/rules/L034.py:99]: Should you have a `set_expression` test case?

File: src/sqlfluff/rules/L034.py [hunk 3]
```diff
@@ -94,10 +94,24 @@ def _eval(self, context: RuleContext) -> EvalResultType:
...
+        ):
+            return None
         if len(context.parent_stack) >= 3 and context.parent_stack[-3].is_type(
             "create_table_statement", "merge_statement"
         ):
             return None
+        if (
+            len(context.parent_stack) >= 4
+            and context.parent_stack[-4].is_type(
+                "create_table_statement", "merge_statement"
```
Comment (tunetheweb) [inline src/sqlfluff/rules/L034.py:110]: Ditto, with `merge_statement` test case?

**Review by tunetheweb — APPROVED** [review]
Approved, with two small suggestions for improvements if you wanna take them on.

File: src/sqlfluff/rules/L034.py (hunk 2 above)
Comment (barrywhart) [reply src/sqlfluff/rules/L034.py:99]: I tried to create one, but couldn't get it to parse.

File: src/sqlfluff/rules/L034.py (hunk 3 above)
Comment (barrywhart) [reply src/sqlfluff/rules/L034.py:110]: Adding one now

**Pull request merged**